### ISSUE TITLE
src/ssl_sock.c: possible resource leak found by coverity

### ISSUE THREAD

**Issue by chipitsine** [haproxy/haproxy#298]
```
    37. Breaking from loop.
    CID 1403834 (#1 of 1): Resource leak (RESOURCE_LEAK)38. leaked_storage: Variable ssl_conf going out of scope leaks the storage it points to.
3874                        break;
3875                }
3876        }
```

**Comment by chipitsine** [comment]
I'm not going to pollute issue tracker with numerous issues :) say, 1-2 issue per week

**Comment by capflam** [comment]
Honestly I'm a bit lost when I read the SSL code. But it seems the `ssl_conf` variable may be referenced in sni contexts when `ssl_sock_add_cert_sni()` is called. It may also be used as default ssl config for a bind config. So here it is hard to know if we can safely release it or not.

That's said, when an error occurred during the configuration parsing, at any place, we just exit. No deinit is performed at all. That makes Valgrind lovers crying. But the config parser is too complex to do such deinit reliably. And it is too much work to handle it. So if it seems to be safe to release it here. But I'm not comfortable with the idea to release something still referenced. It will lead to a crash sooner or later.

So, I'm closing the issue now because there is nothing simple to do to make Coverity happy. And the bug is more general, there is no deinit when we exit on error (if we want to see it as a bug...).

**Comment by chipitsine** [comment]
maybe something like "backlog" and "help wanted" for such issues (instead of closing)

**Comment by chipitsine** [comment]
as for valgrind, we can supply so called "suppression" rules for it

**Comment by capflam** [comment]
Well, I closed the issue because, when a fatal error is reported during the configuration parsing, many variables are not released because it is hard to really know the current state and it does not really matter because HAProxy will exit. But, if you prefer to keep this one open with a specific label, I'm ok.

**Comment by capflam** [comment]
> I'm not going to pollute issue tracker with numerous issues :) say, 1-2 issue per week

thanks for this kind thought Ilya :)

**Comment by chipitsine** [comment]
labels like "backlog" do not pollute tracker much. everyone understands that nobody is going to do anything to those tickets ... unless someone really wants to

**Comment by capflam** [comment]
why not. I will edit labels to add this one.

**Comment by wlallemand** [comment]
the malloc'ed ssl_bind_conf is only used in the case of the crt-list where a ssl_bind_conf is allocated for each line in the crt-list. In the other cases it uses the ssl_bind_conf which is in the bind_conf structure and this pointer remains to NULL in the sni_ctx.

It's difficult to free it directly from the sni_ctx because this pointer can be used by several sni_ctx and there is no refcoun. We can probably add a flag in the first sni_ctx which uses it and only free these ones.

**Comment by chipitsine** [comment]
> We can probably add a flag in the first sni_ctx which uses it and only free these ones.

looks like reimplementing memory management ourselves

**Comment by wlallemand** [comment]
not really, this is just a way to know what to free to avoid a double free.
